### PR TITLE
refactor(sozo): fetch `writes` from events and calculate diff for that + some refactor

### PR DESCRIPTION
**Pull request by lambda-0x**
fix: #2079 

## Summary by CodeRabbit

- **New Features**
  - Introduced a new function for extracting tags from filenames, enhancing filename parsing capabilities.
  - Added a new authorization function that consolidates granting and revoking permissions.
  - Updated migration functions to include additional context and improve clarity during the migration process.
  - Introduced a new function to find authorization differences during migrations, improving resource access management.

- **Bug Fixes**
  - Improved error handling in migration processes to ensure proper validation of changes.

- **Refactor**
  - Streamlined functions to reduce mutability and improve performance by passing parameters by reference.
  - Simplified logic in various asynchronous functions to enhance code clarity and efficiency.

- **Documentation**
  - Enhanced documentation for several functions, clarifying parameters and return types to improve user understanding.

- **Style**
  - Made formatting adjustments across multiple files for consistency and readability.

### CODE REVIEW

File: crates/dojo-world/src/manifest/mod.rs [hunk 1]
```diff
@@ -442,6 +444,23 @@ impl DeploymentManifest {
...
+
+//     pub fn write_to_path_toml(&self, path: &Utf8PathBuf) -> Result<()> {
+//         fs::create_dir_all(path.parent().unwrap())?;
+
+//         let deployed_manifest = toml::to_string_pretty(&self)?;
+//         fs::write(path, deployed_manifest)?;
+
+//         Ok(())
+//     }
+// }
```
Comment (glihm) [inline crates/dojo-world/src/manifest/mod.rs:462]: This is something we may re-use? Or can be removed?

File: crates/dojo-world/src/manifest/types.rs [hunk 2]
```diff
@@ -32,10 +32,19 @@ pub struct BaseManifest {
...
     pub contracts: Vec<Manifest<DojoContract>>,
     pub models: Vec<Manifest<DojoModel>>,
 }
 
+// bool represents authorization has been done for that contract
+// #[derive(Default, Clone, Debug, Serialize, Deserialize)]
+// #[cfg_attr(test, derive(PartialEq))]
+// pub struct DeploymentMetadata {
+//     pub world_metadata: bool,
+// }
```
Comment (glihm) [inline crates/dojo-world/src/manifest/types.rs:46]: Same here if it's something old, we can remove it. 👍

File: crates/dojo-world/src/migration/strategy.rs [hunk 3]
```diff
@@ -145,7 +138,16 @@ pub fn prepare_for_migration(
         world.contract_address = generated_world_address;
     }
 
-    Ok(MigrationStrategy { world_address, world, base, contracts, models, metadata })
+    // If world address is not provided, then we expect the world to be migrated.
+    // TODO: see if there are any case where world_address is not provided and `world` is none
```
Comment (glihm) [inline crates/dojo-world/src/migration/strategy.rs:142]: We should always be able to compute the world address from the local information. We should ensure this is something true all along the migration process.

File: crates/sozo/ops/src/auth.rs [hunk 4]
```diff
@@ -20,6 +20,8 @@ pub enum ResourceType {
     Contract(String),
     Namespace(String),
     Model(String),
+    // this can be a selector for any other resource type
+    Selector(Felt),
```
Comment (glihm) [inline crates/sozo/ops/src/auth.rs:24]: That's nice to have that, good addition.

File: crates/sozo/ops/src/auth.rs [hunk 5]
```diff
@@ -101,7 +107,7 @@ impl FromStr for ResourceOwner {
 pub async fn grant_writer<'a, A>(
     ui: &'a Ui,
     world: &WorldContract<A>,
-    new_writers: Vec<ResourceWriter>,
+    new_writers: &Vec<ResourceWriter>,
```
Comment (glihm) [inline crates/sozo/ops/src/auth.rs:110]: Let's use slices instead of `&Vec`.

File: crates/sozo/ops/src/auth.rs [hunk 6]
```diff
@@ -115,7 +121,7 @@ where
         let resource_selector =
             get_resource_selector(ui, world, &new_writer.resource, default_namespace).await?;
         let contract_address =
-            utils::get_contract_address(world, new_writer.tag_or_address).await?;
+            utils::get_contract_address(world, new_writer.tag_or_address.clone()).await?;
```
Comment (glihm) [inline crates/sozo/ops/src/auth.rs:124]: `get_contract_address` should take a `&str` instead, can be done later though.

File: crates/sozo/ops/src/auth.rs [hunk 7]
```diff
@@ -143,7 +149,7 @@ where
 pub async fn grant_owner<A>(
     ui: &Ui,
     world: &WorldContract<A>,
-    new_owners: Vec<ResourceOwner>,
+    new_owners: &Vec<ResourceOwner>,
```
Comment (glihm) [inline crates/sozo/ops/src/auth.rs:152]: Slice instead of `&Vec`.

File: crates/sozo/ops/src/auth.rs [hunk 8]
```diff
@@ -154,7 +160,8 @@ where
 
     for new_owner in new_owners {
         let resource_selector =
-            get_resource_selector(ui, world, &new_owner.resource, default_namespace).await?;
+            get_resource_selector(ui, world, &new_owner.resource.clone(), default_namespace)
```
Comment (glihm) [inline crates/sozo/ops/src/auth.rs:163]: We shouldn't need to clone if we pass a ref.

```suggestion
            get_resource_selector(ui, world, &new_owner.resource, default_namespace)
```

File: crates/sozo/ops/src/auth.rs [hunk 9]
```diff
@@ -180,7 +187,7 @@ where
 pub async fn revoke_writer<A>(
     ui: &Ui,
     world: &WorldContract<A>,
-    new_writers: Vec<ResourceWriter>,
+    new_writers: &Vec<ResourceWriter>,
```
Comment (glihm) [inline crates/sozo/ops/src/auth.rs:190]: Slice.

File: crates/sozo/ops/src/migration/migrate.rs [hunk 10]
```diff
@@ -882,6 +874,89 @@ pub async fn update_manifests_and_abis(
     Ok(())
 }
 
+pub async fn find_authorization_diff<A>(
```
Comment (glihm) [inline crates/sozo/ops/src/migration/migrate.rs:877]: If you could add a comment that for now we focus on writers, and owners will be added later during a refactor that could be great to not forget it. 👍

File: crates/sozo/ops/src/migration/migrate.rs [hunk 11]
```diff
@@ -882,6 +874,89 @@ pub async fn update_manifests_and_abis(
...
+    let mut grant = vec![];
+    let mut revoke = vec![];
+
+    for c in &diff.contracts {
+        // remote is none meants its not deployed.
+        // now it could have been deployed during this run
+        // which can be checked from migration_output
+        // if c.remote_class_hash.is_none() {
+        //     continue;
+        // }
```
Comment (glihm) [inline crates/sozo/ops/src/migration/migrate.rs:897]: If the check is no longer relevant, we may remove the check code, and only specify why we iterate without checking the remote class hash.

File: crates/sozo/ops/src/migration/mod.rs [hunk 12]
```diff
@@ -44,6 +44,9 @@ pub struct ContractMigrationOutput {
     pub base_class_hash: Felt,
 }
 
+// TODO: read deployment manifest to find diff in `writes` for auto authorize
```
Comment (glihm) [inline crates/sozo/ops/src/migration/mod.rs:47]: Do we need to read deployment manifest if we are fetching events from the chain?

**Review by glihm — COMMENTED** [review]
Nice work mate and good addition of the selector on the auth part. Some changes have been merged on main related to auth, you may have conflict there.

About the writers, it looks good on the logic, and make the flow easy for the user to simply use `migrate apply` and everything is computed.

The only downside of the current implementation is that when revoking, only the selectors are displayed. And fetching each one of the selector info to display the name might be worthy for easier debugging / understanding of the operations, wdyt?

File: crates/dojo-world/src/migration/strategy.rs (hunk 3 above)
Comment (lambda-0x) [reply crates/dojo-world/src/migration/strategy.rs:142]: yup make sense, i just put a TODO there to verify this constraint. But whenever  world_address is not provided we try to deploy it. will remove the TODO.

File: crates/sozo/ops/src/auth.rs (hunk 6 above)
Comment (lambda-0x) [reply crates/sozo/ops/src/auth.rs:124]: will do in this PR

File: crates/sozo/ops/src/migration/mod.rs (hunk 12 above)
Comment (lambda-0x) [reply crates/sozo/ops/src/migration/mod.rs:47]: nope, was stale comment, updated.

**Comment by lambda-0x** [comment]
> And fetching each one of the selector info to display the name might be worthy for easier debugging / understanding of the operations

good point, updated it to show more appropriate resource type instead of selector where possible.

**Review by glihm — APPROVED** [review]

**Pull request merged**